### ISSUE TITLE
[bug] out tree build fails after removing subproject sources directory

### ISSUE THREAD

**Issue by pioneer19** [build2/build2#106]
With out_root != src_root build type after removing subproject, rebuild fails. I expected rebuild will remove subproject in build directory tree mirror. How to reproduce:
```
# first prepare (build project out_root)
bdep new -l c++ -t lib libhello
cd libhello/
bdep init -C ../build-gcc @gcc cc config.cxx=g++
b -sv

# project built, now remove subproject and rebuild
rm -rf tests/
b -sv
```
This will produce: `error: no build/bootstrap.build in tests/`.

I expected `b` will remove `tests/` in `$out_root` and build project.

**Comment by boris-kolpackov** [comment]
Yes, this "stray output without any trace of it in source" situations are always tricky since we cannot be sure whether it is indeed the output that we have created (most likely) or it's something that was created by the user (less likely but possible). And under this uncertainty it's probably not wise to always assume the former and start deleting things.

The two possible solutions that I can see are to either issue better diagnostics, perhaps suggesting that the user deletes the stray
directory, or to simply ignore it. The ignore case, in fact, would be consistent with the case where `tests/` isn't a subproject: after deleting it in the source directly we would have the corresponding output (object files, executables, etc) silently lying around until removed manually by the user (e.g., as part of dropping the configuration).

**Comment by Klaim** [comment]
I didn't realize that it was working like this but it explains some of the issues I saw in the past.

> after deleting it in the source directly we would have the corresponding output (object files, executables, etc) silently lying around until removed manually by the user (e.g., as part of dropping the configuration)

Are they truly silent? Isn't there ways where the build would be influenced by the presence of these unused directories? 

I'm wondering if a command/option to "strictly synchronize" outdir would be helpful, in that it would clarify to `build2` the user's intention that the outdir should reflect strictly the sourcedir, so `build2` should remove any unused files. Wouldn't that help?

I'm for the additional diagnostic.

**Comment by boris-kolpackov** [comment]
> Are they truly silent? Isn't there ways where the build would be influenced by the presence of these unused directories?

Theoretically, yes. For example, you could have a generated header that you've removed from the build and then added a non-generated header with the same name. But that feels quite far-fetched and also it's unlikely to go unnoticed.

> I'm wondering if a command/option to "strictly synchronize" outdir would be helpful, in that it would clarify to build2 the user's intention that the outdir should reflect strictly the sourcedir, so build2 should remove any unused files. Wouldn't that help?

It won't be easy to implement the "unused" check (the only proper way would be to dry-run every operation on every target and then notice files that haven't been produced by any of them). Perhaps that `bdep-reset` idea that keeps coming up could be a solution to this problem as well. That is, we would completely clean the configuration of all output and then re-init every package.

**Comment by boris-kolpackov** [comment]
Adding some more notes from the slack chat:

> I think (but not 100% certain) that you should be able to do this by manually removing the subproject and then doing explict `bdep sync --all --disfigure`. But that will also drop any manual configuration you may have for the packages involved. Currently the thinking is that `bdep-reset` will be the command that will be able to handle this correctly.
